PLAT-98205: Separate logics in moonstone/VirtualListBase, ScrollerBase, and ScrollableBase(Native) into custom Hooks

ybsung opened the pull request:
This PR will be merged after the PR - https://github.com/enactjs/enact/pull/2670 is merged and this PR refreshes.

### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

There were so many functions defined in moonstone/VirtualListBase, ScrollerBase, and ScrollableBase(Native). It would be better to separate those functions into custom hooks to enhance maintenance and readability.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

Created proper custom hooks.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

There were duplicated logics in moonstone/Scrollable and moonstone/ScrollableNative. By using custom Hooks, I could reduce the duplicated logics. Now two components are quite similar. We could unify them into one in PLAT-98908 easily.

### Links
[//]: # (Related issues, references)

PLAT-98205

### Comments

1. There are useSpottable hooks in moonstone/VirtualList, moonstone/Scrollable, and moonstone/Scroller. We might need to unify similar hooks in those components into one hook in another issue in future.
2. `useCallback` will be applied later in PLAT-98958 because this PR already includes too much.

MikyungKim commented on packages/moonstone/Scrollable/useSpottable.js:
```diff
@@ -0,0 +1,243 @@
+	});
+
+	/*
+	 * useEffects
```
Why don't we replace it with a proper name?

MikyungKim commented on packages/moonstone/VirtualList/useEventKey.js:
```diff
@@ -0,0 +1,247 @@
+	const {spottable: {current: {uiRefCurrent}}} = instances;
+	const {
+		containerNode,
+		handlerGlobalKeyDownCB,
```
I assume handleGlobalKeyDownCB ?

ybsung commented:
@MikyungKim 

> Check your indentation option.

Fixed

> maybe... handleGlobalKeyDownCB?

Now removed

> I think we need to import setPointerMode if you want to pass this method. In fact, I'm wondering why we should pass spotlight related methods to useKey.

Passed SpotlightAccelerator, then SpotlightAccelerator.setPointerMode() will be called.

> Please return setContainerDisabled too.
> and also calculatePositionOnFocus.

Fixed

> Why don't we replace it with a proper name?

Fixed  to `* Hooks`

> I assume handleGlobalKeyDownCB ?

Now removed

MikyungKim commented on packages/moonstone/Scrollable/useOverscrollEffect.js:
```diff
@@ -0,0 +1,109 @@
+	 * Instance
+	 */
+
+	var variables = useRef({
```
var -> const

MikyungKim commented on packages/moonstone/Scrollable/useOverscrollEffect.js:
```diff
@@ -0,0 +1,109 @@
+
+	function clearOverscrollEffect (orientation, edge) {
+		variables.current.overscrollJobs[orientation][edge].startAfter(overscrollTimeout, orientation, edge, overscrollTypeNone, 0);
+		uiRef.current.setOverscrollStatus(orientation, edge, overscrollTypeNone, 0);
```
Where is uiRef?

MikyungKim commented on packages/moonstone/Scrollable/useEventWheel.js:
```diff
@@ -0,0 +1,146 @@
+
+				if (!variables.current.isWheeling) {
+					if (!props['data-spotlight-container-disabled']) {
+						childRef.current.setContainerDisabled(true);
```
We may remove childRef.current.

MikyungKim commented on packages/moonstone/Scrollable/useEventMouse.js:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	function handleMouseDown (ev) {
+		if (isScrollButtonFocused()) {
```
Can't find this function.

MikyungKim commented on packages/moonstone/Scrollable/useEventVoice.js:
```diff
@@ -0,0 +1,136 @@
+
+	const {direction} = props;
+	const {uiRef} = instances;
+	const {onScrollbarButtonClick} = dependencies
```
missed ;

MikyungKim commented on packages/moonstone/Scrollable/useEventVoice.js:
```diff
@@ -0,0 +1,136 @@
+
+	const variables = useRef({
+		isVoiceControl: false,
+		voiceControlDirection: 'vertical',
```
additional ,

MikyungKim commented on packages/moonstone/Scroller/useSpotlightConfig.js:
```diff
@@ -0,0 +1,55 @@
+	 */
+
+	function handleLeaveContainer ({direction, target}) {
+		const {current: {containerRef, scrollBounds: {maxLeft, maxTop}, scrollPos: {left, top}}} = uiRef.current;
```
suspicious current: part.

MikyungKim commented on packages/moonstone/Scroller/useSpottable.js:
```diff
@@ -0,0 +1,262 @@
+
+	useEffect(() => {
+		return () => setContainerDisabled(false);
+	}, [setContainerDisabled]);	// TODO : Handle exhaustive-deps ESLint rule.
```
I think it depends on `spotlightId`, not `setContainerDisabled`.

0x64 commented:
LGTM!
